### ISSUE TITLE
NoMethodError upon initialize

### ISSUE THREAD

**Issue by jborrey** [twitter/secureheaders#299]
### Expected outcome

`SecureHeaders::Configuration.default` should not crash.

### Actual outcome

`SecureHeaders::Configuration.default` crashes.

In a ruby terminal
```ruby
require 'secure_headers'
SecureHeaders::Configuration.default
```

```
NoMethodError: undefined method `>=' for nil:NilClass
lib/secure_headers/headers/content_security_policy.rb:217:in `supported_directives'
```

Stacktrace:

```
secure_headers-3.5.0/lib/secure_headers/headers/content_security_policy.rb:217:in `supported_directives'
secure_headers-3.5.0/lib/secure_headers/headers/content_security_policy.rb:65:in `normalize_child_frame_src'
secure_headers-3.5.0/lib/secure_headers/headers/content_security_policy.rb:31:in `initialize'
secure_headers-3.5.0/lib/secure_headers/headers/policy_management.rb:192:in `new'
secure_headers-3.5.0/lib/secure_headers/headers/policy_management.rb:192:in `make_header'
secure_headers-3.5.0/lib/secure_headers/configuration.rb:297:in `block in generate_csp_headers_for_config'
secure_headers-3.5.0/lib/secure_headers/configuration.rb:296:in `each'
secure_headers-3.5.0/lib/secure_headers/configuration.rb:296:in `generate_csp_headers_for_config'
secure_headers-3.5.0/lib/secure_headers/configuration.rb:289:in `generate_csp_headers'
secure_headers-3.5.0/lib/secure_headers/configuration.rb:276:in `cache_headers!'
secure_headers-3.5.0/lib/secure_headers/configuration.rb:74:in `add_configuration'
secure_headers-3.5.0/lib/secure_headers/configuration.rb:18:in `default'
```

Am I missing something in configuration?

**Comment by oreoshake** [comment]
@jborrey Wow. So fail, but I can't reproduce. And there's lots of tests using the same code: https://github.com/twitter/secureheaders/search?utf8=%E2%9C%93&q=Configuration.default

What version of ruby are you using? 

```
[Neils-MacBook-Pro secureheaders (master)]$ gem install secure_headers -v 3.5.0
Fetching: secure_headers-3.5.0.gem (100%)
Successfully installed secure_headers-3.5.0
Parsing documentation for secure_headers-3.5.0
Installing ri documentation for secure_headers-3.5.0
Done installing documentation for secure_headers after 0 seconds
1 gem installed
[Neils-MacBook-Pro secureheaders (master)]$ irb
irb(main):001:0> require 'secure_headers'
=> true
irb(main):002:0> SecureHeaders::Configuration.default
=> #<SecureHeaders::Configuration:0x007fea27159488 @hpkp=#<SecureHeaders::NoOpHeaderConfig:0x007fea27825eb0>, @referrer_policy=#<SecureHeaders::NoOpHeaderConfig:0x007fea27825eb0>, @csp=#<SecureHeaders::ContentSecurityPolicyConfig:0x007fea271586c8 @default_src=["'self'", "https:"], @modified=false, @font_src=["'self'", "https:", "data:"], @img_src=["'self'", "https:", "data:"], @object_src=["'none'"], @script_src=["https:"], @style_src=["'self'", "https:", "'unsafe-inline'"]>, @csp_report_only=#<SecureHeaders::NoOpHeaderConfig:0x007fea27825eb0>, @cached_headers={:hsts=>["Strict-Transport-Security", "max-age=631138519"], :x_content_type_options=>["X-Content-Type-Options", "nosniff"], :x_download_options=>["X-Download-Options", "noopen"], :x_frame_options=>["X-Frame-Options", "sameorigin"], :x_permitted_cross_domain_policies=>["X-Permitted-Cross-Domain-Policies", "none"], :x_xss_protection=>["X-XSS-Protection", "1; mode=block"], :csp=>{"Chrome"=>["Content-Security-Policy", "default-src 'self' https:; font-src 'self' https: data:; img-src 'self' https: data:; object-src 'none'; script-src https:; style-src 'self' https: 'unsafe-inline'"], "Opera"=>["Content-Security-Policy", "default-src 'self' https:; font-src 'self' https: data:; img-src 'self' https: data:; object-src 'none'; script-src https:; style-src 'self' https: 'unsafe-inline'"], "Firefox"=>["Content-Security-Policy", "default-src 'self' https:; font-src 'self' https: data:; img-src 'self' https: data:; object-src 'none'; script-src https:; style-src 'self' https: 'unsafe-inline'"], "FirefoxTransitional"=>["Content-Security-Policy", "default-src 'self' https:; font-src 'self' https: data:; img-src 'self' https: data:; object-src 'none'; script-src https:; style-src 'self' https: 'unsafe-inline'"], "Safari"=>["Content-Security-Policy", "default-src 'self' https:; font-src 'self' https: data:; img-src 'self' https: data:; object-src 'none'; script-src https:; style-src 'self' https: 'unsafe-inline'"], "Edge"=>["Content-Security-Policy", "default-src 'self' https:; font-src 'self' https: data:; img-src 'self' https: data:; object-src 'none'; script-src https:; style-src 'self' https: 'unsafe-inline'"], "Other"=>["Content-Security-Policy", "default-src 'self' https:; font-src 'self' https: data:; img-src 'self' https: data:; object-src 'none'; script-src https:; style-src 'self' https: 'unsafe-inline'"]}}, @hpkp_report_host=nil>
irb(main):004:0> SecureHeaders::Configuration.method(:default).source_location
=> ["/Users/neil/.rbenv/versions/2.2.5/lib/ruby/gems/2.2.0/gems/secure_headers-3.5.0/lib/secure_headers/configuration.rb", 15]
```

**Comment by jborrey** [comment]
Yeah I assumed something must be up since this is quite a critical code path and everyone else seems to have this working.

```
ruby -v
ruby 2.3.1p112 (2016-04-26 revision 54768) [x86_64-darwin15]
```

Trying this in `irb` version `0.9.6(09/06/30)`

**Comment by oreoshake** [comment]
This is very scary. Exact version match, still can't repro. 

```
[Neils-MacBook-Pro secureheaders (master *)]$ irb -v
irb 0.9.6(09/06/30)
[Neils-MacBook-Pro secureheaders (master *)]$ ruby -v
ruby 2.3.1p112 (2016-04-26 revision 54768) [x86_64-darwin15]
[Neils-MacBook-Pro secureheaders (master *)]$ irb
irb(main):001:0> require 'secure_headers'
=> true
irb(main):002:0> SecureHeaders::Configuration.default
=> #<SecureHeaders::Configuration:0x007fb768a56000 @hpkp=#<SecureHeaders::NoOpHeaderConfig:0x007fb768a5f290>, @referrer_policy=#<SecureHeaders::NoOpHeaderConfig:0x007fb768a5f290>, @csp=#<SecureHeaders::ContentSecurityPolicyConfig:0x007fb768a54d68 @default_src=["'self'", "https:"], @modified=false, @font_src=["'self'", "https:", "data:"], @img_src=["'self'", "https:", "data:"], @object_src=["'none'"], @script_src=["https:"], @style_src=["'self'", "https:", "'unsafe-inline'"]>, @csp_report_only=#<SecureHeaders::NoOpHeaderConfig:0x007fb768a5f290>, @cached_headers={:hsts=>["Strict-Transport-Security", "max-age=631138519"], :x_content_type_options=>["X-Content-Type-Options", "nosniff"], :x_download_options=>["X-Download-Options", "noopen"], :x_frame_options=>["X-Frame-Options", "sameorigin"], :x_permitted_cross_domain_policies=>["X-Permitted-Cross-Domain-Policies", "none"], :x_xss_protection=>["X-XSS-Protection", "1; mode=block"], :csp=>{"Chrome"=>["Content-Security-Policy", "default-src 'self' https:; font-src 'self' https: data:; img-src 'self' https: data:; object-src 'none'; script-src https:; style-src 'self' https: 'unsafe-inline'"], "Opera"=>["Content-Security-Policy", "default-src 'self' https:; font-src 'self' https: data:; img-src 'self' https: data:; object-src 'none'; script-src https:; style-src 'self' https: 'unsafe-inline'"], "Firefox"=>["Content-Security-Policy", "default-src 'self' https:; font-src 'self' https: data:; img-src 'self' https: data:; object-src 'none'; script-src https:; style-src 'self' https: 'unsafe-inline'"], "FirefoxTransitional"=>["Content-Security-Policy", "default-src 'self' https:; font-src 'self' https: data:; img-src 'self' https: data:; object-src 'none'; script-src https:; style-src 'self' https: 'unsafe-inline'"], "Safari"=>["Content-Security-Policy", "default-src 'self' https:; font-src 'self' https: data:; img-src 'self' https: data:; object-src 'none'; script-src https:; style-src 'self' https: 'unsafe-inline'"], "Edge"=>["Content-Security-Policy", "default-src 'self' https:; font-src 'self' https: data:; img-src 'self' https: data:; object-src 'none'; script-src https:; style-src 'self' https: 'unsafe-inline'"], "Other"=>["Content-Security-Policy", "default-src 'self' https:; font-src 'self' https: data:; img-src 'self' https: data:; object-src 'none'; script-src https:; style-src 'self' https: 'unsafe-inline'"]}}, @hpkp_report_host=nil>
irb(main):003:0>
```

**Comment by jborrey** [comment]
Figured this out. My version of the `useragent` gem was `0.10.0` which does not populate the `@version` field on the `UserAgent` object. So you would get the `NoMethodError` when operating on `nil` later on in the code.

**Comment by oreoshake** [comment]
Aha! Well it's my bug, it should be resilient to that field and others being nil. I can take a look in the upcoming weeks unless someone else wants to cut a PR :smile:

**Comment by jborrey** [comment]
So you would prefer the fix to be just making resilience to `nil` instead of pinning the gem version? I can probably do that if you prefer it.

**Comment by oreoshake** [comment]
> So you would prefer the fix to be just making resilience to nil instead of pinning the gem version?

Absolutely :100:. Libraries should avoid pinning at all costs IMO 👍

**Comment by stve** [comment]
@oreoshake I think we can close this and #300 now that #301 has been merged

**Comment by jborrey** [comment]
Yes I will close it. Thanks for getting a PR out for that!

**Comment by oreoshake** [comment]
Released in 3.5.1
